src/Kriti/Parser/Monad.hs

```diff
@@ -0,0 +1,216 @@
+        _span = s
+      }
+  render (UnexpectedToken tok) =
+    let tok' = serialize $ unLoc tok
+        span' = locate tok
+     in E.RenderedError
+          { _code = E.ParseErrorCode,
+            _message = "ParseError: Unexpected token '" <> tok' <> "'.",
+            _span = span'
+          }
+  render (InvalidLexeme start inp) =
+    E.RenderedError
+      { _code = E.LexErrorCode,
+        _message = "LexError: Invalid Lexeme: '" <> TE.decodeUtf8 inp <> "'",
+        _span = Span start (overCol (+ (B.length inp)) start)
+      }
+
+parseError :: ParseError -> Parser a
+parseError err = throwError err
+
+--------------
+--- Tokens ---
+--------------
+
+-- | Construct a Token from the matched Text and the current `parseSpan`.
+{-# INLINE token #-}
+token :: (Loc T.Text -> Token) -> B.ByteString -> Parser Token
+token k bs = k <$> located (TE.decodeUtf8 bs)
+
+-- | Construct a `(TokenSymbol (Loc _))` using the current `parseSpan`
+-- to construct the `Loc _`.
+{-# INLINE symbol #-}
+symbol :: Symbol -> B.ByteString -> Parser Token
+symbol sym _ = do
+  sp <- location
+  pure $ TokSymbol $ Loc sp sym
+
+-----------------------
+--- Alex Primitives ---
+-----------------------
+
+data AlexInput = AlexInput
+  { lexPos :: AlexSourcePos,
+    lexPrevChar :: Char,
+    -- | current input bytestring
+    lexBytes :: B.ByteString,
+    -- | remaining bytes in current character
+    lexCharBytes :: [Word8]
+  }
+
+{-# INLINE nextLine #-}
+nextLine :: B.ByteString -> AlexInput -> AlexInput
+nextLine rest AlexInput {..} =
+  AlexInput
+    { lexPos = lexPos {line = line lexPos + 1, col = 1},
+      lexPrevChar = '\n',
+      lexBytes = rest,
+      lexCharBytes = []
+    }
+
+{-# INLINE nextCol #-}
+nextCol :: Char -> B.ByteString -> AlexInput -> AlexInput
+nextCol c rest AlexInput {..} =
+  AlexInput
+    { lexPos = lexPos {col = col lexPos + 1},
+      lexPrevChar = c,
+      lexBytes = rest,
+      ..
+    }
+
+{-# INLINE popBufferedBytes #-}
+popBufferedBytes :: AlexInput -> Maybe (Word8, AlexInput)
+popBufferedBytes AlexInput {..} =
+  case lexCharBytes of
+    [] -> Nothing
+    (b : bs) -> Just (b, AlexInput {lexCharBytes = bs, ..})
+
+{-# INLINE bufferBytes #-}
+bufferBytes :: Char -> [Word8] -> B.ByteString -> AlexInput -> AlexInput
+bufferBytes c bytes rest AlexInput {..} =
+  AlexInput
+    { lexPrevChar = c,
+      lexBytes = rest,
+      lexCharBytes = bytes,
+      ..
+    }
+
+alexGetByte :: AlexInput -> Maybe (Word8, AlexInput)
+alexGetByte input@AlexInput {..} =
+  case popBufferedBytes input of
+    Nothing -> advance' <$> UTFBS.uncons lexBytes
+    ok -> ok
+  where
+    advance' :: (Char, B.ByteString) -> (Word8, AlexInput)
+    advance' ('\n', rest) = (B.c2w '\n', nextLine rest input)
+    advance' (c, rest) =
+      case UTF8.encodeChar c of
+        [b] -> (b, nextCol c rest input)
+        (b : bs) -> (b, bufferBytes c bs rest input)
+        [] -> error "The impossible happened! A Char decoded to 0 bytes."
```

Comment:
It looks like this is an `encoding` failure, not a `decoding` failure; additionally, the error message should include the aberrant character.

For example:
`error $ "The following character produced an empty UTF8-encoded bytestring, which should be impossible: " ++ [ c ]`

Reply:
good catch!